File: 44-wildcard-matching/wildcard-matching.py
from functools import cache

class Solution:
    def isMatch(self, s: str, t: str) -> bool:
        @cache
        def match(si: int, ti: int) -> bool:
            if si < 0 and ti < 0:
                return True
            if ti < 0:
                return False
            if t[ti] == '*':
                if match(si, ti - 1):
                    return True
                if si >= 0 and match(si - 1, ti):
                    return True
                if si >= 0 and match(si - 1, ti - 1):
                    return True
                return False
            if si < 0:
                return False
            if s[si] == t[ti] or t[ti] == '?':
                return match(si - 1, ti - 1)
            return False

        return match(len(s) - 1, len(t) - 1)

File: 44-wildcard-matching/wildcard-matching.cpp
class Solution {
public:
    bool isMatch(string s, string t) {
        if(s.empty()){
            for(char ch:t) if(ch!='*') return 0;
            return 1;
        }
        vector<vector<bool>> dp(s.size()+1,vector<bool>(t.size()+1,0));
        dp[s.size()][t.size()]=1;
        for(int j=t.size()-1;j>=0;j--){
            dp[s.size()][j]=(t[j]=='*' && dp[s.size()][j+1]);
        }
        for(int i=s.size()-1;i>=0;i--){
            for(int j=t.size()-1;j>=0;j--){
                dp[i][j]=((s[i]==t[j] || t[j]=='?') && dp[i+1][j+1])
                        ||(t[j]=='*' && (dp[i+1][j] || dp[i][j+1]));
            }
        }
        return dp[0][0];
    }
};